File: listas_enlazadas.c
/******************************************************************************

                            Online C Compiler.
                Code, Compile, Run and Debug C program online.
Write your code in this editor and press "Run" button to compile and execute it.

*******************************************************************************/
#ifndef libro_h
#define libro_h

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Libro {
    char titulo[50];
    char autor[50];
    char isbn[13];
    
} Libro;

/*Definimos una estrucutura como nodo que tiene una referencia al Libro*/
typedef struct Nodo{
    Libro libro; // Guarda un libro
    struct Nodo* siguiente;    // Guarda un puntero al siguiente elemento
} Nodo;

// Creamos otra estructura que guarda el puntero a la cabeza de esa lista
typedef struct Lista{
    Nodo* cabeza;
    int longitud; // Dice cuantos elementos tenemos en la lista (Cuando incremente se lo sumamos a esta varaible)(Cuando decremente se lo restamos a esta varaible )
}Lista;

// Crear un nuevo Nodo
Nodo* CrearNodo(libro libro){
   Nodo* nodo = (Nodo*)malloc(sizeof(Nodo)); // Instanciar un nuevo nodo usando la funcion malloc()
   strcpy(nodo->libro.titulo, libro->titulo,50);// Copia dentro del libro los datos de los parametros que se pasan del libro
   strcpy(nodo->libro.autor, libro->autor,50);
   strcpy(nodo->libro.isbn, libro->isbn,50);
   nodo -> siguiente = NULL; // Como no se sabe a que apunta el siguiente lo ponemos a apuntar a null
   return nodo
}

// Destruir un nodo que ya no queramos en memoria
void DestruirNodo(Nodo* nodo){
    free(nodo);
}

// Insertar un libro al principio de una lista
void InsertarPrincipio(Lista* lista, Libro* libro){
    *Nodo nodo = CrearNodo(libro); // Creamos un nuevo nodo a partir de los datos del libro
    nodo->siguiente = lista->cabeza;// Colocamos como cola del nodo que se acaba de crear la actual cabeza de la lista
    lista->cabeza = nodo;// Movemos el puntero de la cabeza de la lista para que apunte al nuevo nodo
    lista->longitud++;
}
// Insertar un libro al final de una lista
void InsertarFinal(Lista* lista, Libro* libro){ // Esta funcion no va a funcionar si la lisa esta vacia
    Nodo* nodo = CrearNodo(libro); //Creamos otro nodo
    if(lista->cabeza == NULL){
        lista->cabeza = nodo; //Se pociciona directamente como cabeza de la lista
    }else{
    Nodo* puntero = lista->cabeza;// Creamos un nodo especial al que va a hacer la cabeza de la lista
    // Hacemos que el puntero vaya de una pocision a otra hasta que alcnace el final de la lista
    while (puntero->siguiente){
        puntero = puntero->siguiente; // El puntero avanza en una pocision 
    }
        
    }
    // Ya sabemos que el puntero en el que estamos no tiene otro siguiente por ende ponemos el nodo que queremos insertar como el siguiente del nodo actual
    puntero->siguiente = nodo;
    lista->longitud++;
}
void InsertarDespues(int n, Lista* lista, Libro* libro){
    Nodo* nodo = CrearNodo(libro);
    if(lista->cabeza == NULL){// Lista vacia
        lista->cabeza = nodo;
        
    }else{
        // Debemos recorrer la lista hasta que lleguemos a la pocision anterior
        int pocision = 0;
        while(pocision < n && puntero->siguiente){
            puntero = puntero->siguiente;
            pocision++; //Cada vez que se recorra este contador aumenta
        }
        // Cuando tenemos el libro n en la pocision del puntero, debemos insertarlo despues
        nodo->siguiente = puntero->siguiente; //Conectamos por la derecha
        puntero->siguiente = nodo;// Conectamos por izquierda
        

    }
    lista->longitud++;
}

// Funcion que nos devuelve el libro que este en la pocision n de la lista 
Libro* Obtener(int n, Lista* lista){
    // Verificamos si la lista esta vacia
    if(lista->cabeza == NULL){
        return NULL
    }else{
        // Se hace algo parecido con la funcion anterior
        Nodo* puntero = lista->cabeza;
        int pocision = 0;
        while (pocision < n && puntero ->siguiente){
            puntero = puntero->siguiente; // Avanza hasta el elemento n
            pocision++;
        }
        
        // Cuando ya esta en el elemento n retorna el libro de esta pocision
        if (pocision != n){ // Este condicional lo que hace es retornar null si la pocision no es la misma que n por que n es mayor
            return NULL;
        }else{
        return &puntero->libro;}
    }
}
// Saber cuantos elementos tengo en la lista
int Contar(Lista* lista){
    return lista->longitud;
}

// ELIMINAR ELEMENTOS DE LA LISTA
int EstaVacia(Lista* lista){ //Determina si la lista esta vacia
    return lista->cabeza == NULL; // Devuelve 0 si la lista esta vacia o 1 si tiene cabeza
}
void EliminarPrincipio(Lista* lista){
    if(lista->cabeza){ // Se ejecuta cuando la lista NO esta vacia
        Nodo* eliminado = lista->cabeza; // Se guarda la cabeza que se va a eliminar
        lista->cabeza = lista->cabeza->siguiente; // El siguiente nodo de la cabeza pasa a hacer la cabeza
        DestruirNodo(eliminado) // Se elimina el nodo que guardamos
        lista->longitud--; //Decrementamos la longitud de la lista
    }
        
}

void EliminarUltimo(Lista* lista){
    if(lista->cabeza){ 
        if(lista->cabeza->siguiente){ //Sirve para cuando la lista tiene mas de un elemento
            Nodo* puntero = lista->cabeza;
            while(puntero->siguiente->siguiente){ //Llegamos al penultimo elemento de la lista cuando el siguiente del siguiente valga nulo 
                puntero = puntero->siguiente;
            
            }
            // Cuando llegamos a la penultima pocision eliminamos el ultimo nodo
            Nodo* elimiando = puntero->siguiente; // Guardamos el nodo que queremos eliminar
            puntero->siguiente = NULL; // Hacemos que el penultimo nodo ahora apunte a null lo que lo convierte en el ultimo
            DestruirNodo(eliminado) //Destruimos el nodo
            lista->longitud--;
    }else{ // Se ejecuta si la lista solo tiene un elemento
        // Se elimina la cabeza de la lista
        Nodo* eliminado = lista->cabeza;
        lista->cabeza = NULL;
        DestruirNodo(eliminado);
        lista->longitud--;
    }
}
}

void EliminarElemento(int n, Lista* lista){
    if(lista->cabeza){
        if(n == 0){ // Si se quiere eliminar el primer elemento
            Nodo* eliminado = lista->cabeza; // Se guarda la cabeza que se va a eliminar
            lista->cabeza = lista->cabeza->siguiente; // El siguiente nodo de la cabeza pasa a hacer la cabeza
            DestruirNodo(eliminado) // Se elimina el nodo que guardamos
            lista->longitud--; //Decrementamos la longitud de la lista
            
        }else if (n<lista->longitud){ // Nos aseguramos de que n sea menor a los libros que tenemos en la lista
            Nodo* puntero = lista->cabeza;
            int pocision = 0;
            while(pocision<(n-1)){ // Recorremos la lista hasta llegar al nodo siguiente
                puntero = puntero->siguiente;
                pocision++;
            }
        
            // Una vez llegamos eliminamos el nodo siguiente
            Nodo* eliminado = puntero->siguiente;
            puntero->siguiente = eliminado->siguiente; // Conectamos el nodo siguiente del anterior con el siguiente que se va a eliminar
            DestruirNodo(elminado);
            lista->longitud--;
    }
    }
}
#endif /* libro_h */

File: LISTAS ENLAZADAS/Listas_enlazadas.java
/******************************************************************************

Welcome to GDB Online.
GDB online is an online compiler and debugger tool for C, C++, Python, Java, PHP, Ruby, Perl,
C#, OCaml, VB, Swift, Pascal, Fortran, Haskell, Objective-C, Assembly, HTML, CSS, JS, SQLite, Prolog.
Code, Compile, Run and Debug online from anywhere in world.

*******************************************************************************/
// Listas enlazadas
// Estructura del libro
public class Libro{
    private String titulo;
    private String autor;
    private String isbn;
    
    public Libro(String titulo, String autor, String isbn){
        this.titulo = titulo;
        this.autor = autor;
        this.isbn = isbn; 
    }
    
    public String getTitulo(){return titulo;}
    public void setTitulo(String titulo){ this.titulo = titulo;}
    public String getAutor(){return autor;}
    public void setAutor(String autor){this.autor = autor;}
    public String getIsbn(){return isbn;}
    public void setIsbn(String isbn){this.isbn = isbn;}
}

public class Lista{
    private Nodo cabeza = null; // Mantiene una referencia a la cabeaz de la lista
    private int longitud = 0;
    // Estructura Interna
    private class Nodo {
        private Libro libro; // Referencia al libro que tenga asociado el nodo
        public Nodo siguiente = null; // Referencia al nodo siguiente de la lista
        // Java elimina automaticamente los nodos que no utilicemos
        public Nodo(Libro libro){
            this.libro = libro;
        }
    }
    
    // Insertar al principio de la lista
    public void insertarPrincipio(Libro libro){
        // Creamos un nuevo Nodo
        Nodo nodo = new Nodo(libro);
        // Hacemos que lo que era la cabeza de la lista se vuelva el nodo siguiente del elemento que acabamos de crear
        nodo.siguiente = cabeza;
        cabeza = nodo;
        longitud++;
    }
    
    // Insertar al final de una lista
    public void insertarFinal(Libro libro){
        // Recorrer una lista
        Nodo nodo = new Nodo(libro); // Nodo que vamos a crear
        Nodo puntero = cabeza // Este puntero al principio va a ir en la cabeza de la lista
        while(puntero.siguiente != null){ // Itera mientras el puntero siguiente no sea nulo
            puntero = puntero.siguiente; // El puntero avanza por toda la lista
        }
        // Cuando ya el siguiente elemento del puntero valga null hacemos que ahora valga el elemento que acabamos de crear
        puntero.siguiente = nodo;
        longitud++;
        
    }
    // Insertar un elemento en la mitad de la lista
    public void insertarDespues(int n, Libro libroi){
        // Creamos el nuevo nodo
        Nodo nodo = new Nodo(libro);
        // Comprobamos que la lista no este vacia
        if (cabeza == null){
            cabeza = nodo
        }else{
            // Recorremos la lista y localizamos el numero n
            Nodo puntero = cabeza // Este puntero al principio va a ir en la cabeza de la lista
            int contador = 0; // Introducimos un contador que nos de la pocision de cada elemento de la lista
            while(contador < n && puntero.siguiente != null){ 
                puntero = puntero.siguiente;
                contador++; // Cada vez que pase a otro puntero el contador aumenta
            }
            
            nodo.siguiente = puntero.siguiente; // Conectamos el nodo recien creado con la lista por la derecha
            puntero.siguiente = nodo; // Conectamos el nodo recien creado por la izquierda
            
        }
        longitud++;
    }
    
    // Obtener informacion de un libro dada una pocision en la lista
    public Libro obtener(int n){
        if (cabeza == null){// Comprobamos si la lista esta vacia
            return null; // No se puede retornar nada si la lista esta vacia
        }else{
            Nodo puntero = cabeza // Este puntero al principio va a ir en la cabeza de la lista
            int contador = 0; // Introducimos un contador que nos de la pocision de cada elemento de la lista
            while(contador < n && puntero.siguiente != null){ 
                puntero = puntero.siguiente;
                contador++; // Cada vez que pase a otro puntero el contador aumenta
            }
            if (contador != n){//Si despues del bucle contador no vale lo que vale n
                return null; // Hemos alcanzado el final de la lista de forma permatura
            }else {
                return puntero.libro; // Retorna el libro que esta en la pocision n
            }
                
        }
            
        }
        
        // Funcion para que nos diga cuantos elementos hay en nuestra lista
        public int contar(){
            return longitud;
        }
        
        // ELIMINAR ELEMENTOS DE LA lista
        // Booleano que muestra si la lista esta vacia o no
        public boolean estaVacia(){
            return cabeza == null;
        }
        // Elimianr primer elemento de la lista
        public void eliminarPrimero(){
            if (cabeza != null){ // Solo podemos eliminar cosas si la lista no esta vacia
                Nodo primer = cabeza; // El primer elemento de la lista es la cabeza
                // Hacemos que la cabea en vez de apuntar al primero apunte al segundo
                cabeza = cabea.siguiente;
                // Hacemos que el nodo que quitamos apunte a null
                primer.siguiente = null;
                longitud--;
            }
        }
        
        // Eliminar un elemento que este al final de la lista
        public void eliminarUltimo(){
            // Debemos recorrer la lista hasta alcanzar el penultimo elemento
            if(cabeza != null){
                // Si la lista solo tiene un elemento
                if (cabeza.siguiente == null){
                    // Eliminamos el unico elemento que queda en la lista
                    cabeza = null;
                }else{
                Nodo puntero = cabeza;
                // Llegaremos al penultimo elemento cuando el siguiente de su siguiente sea null
                while(puntero.siguiente.siguiente != null){
                    puntero = puntero.siguiente;
                }
                // Eliminamos la referencia con el penultimo
                puntero.siguiente = null;
                longitud--;
                } 
            }
        }
        
        // Eliminar un libro en una pocision
        public void eliminarLibro(int n){
            // Comprobamos que la lista no este vacia
            if(cabeza != null){
                // Eliminar el primer libro
                if (n == 0){
                    Nodo primer = cabeza;
                    cabeza = cabeza.siguiente;
                    primer.siguiente = null;
                    longitud--;
                }else (n< longitud){
                    Nodo puntero = cabeza;
                    int contador = 0;
                    while (contador < (n-1)){
                        puntero = puntero.siguiente;
                        contador++;
                    }
                    Nodo temp = puntero.siguiente // Lo marcamos como elemento que queremos descartar
                    // Lo conectamos por izquierda y derecha a la lista
                    puntero.siguiente = temp.siguiente // Marca al siguiente nodo que queremos descartar
                    temp.siguiente = null; // Eliminamos la referencia
                    longitud --;
            }
        }
    }
}

